promql/engine.go

```diff
@@ -1088,6 +1088,146 @@ func (ev *evaluator) rangeEval(funcCall func([]parser.Value, *EvalNodeHelper) (V
 	return mat, warnings
 }
 
+// TODO in this experiment, the checking of ev.currentSamples is currently skipped (but easy to do).
+func (ev *evaluator) rangeEvalSum(grouping []string, without bool, exprs ...parser.Expr) (Matrix, storage.Warnings) {
+	numSteps := int((ev.endTimestamp-ev.startTimestamp)/ev.interval) + 1
+	matrixes := make([]Matrix, len(exprs))
+	origMatrixes := make([]Matrix, len(exprs))
+	originalNumSamples := ev.currentSamples
+
+	var warnings storage.Warnings
+	for i, e := range exprs {
+		// Functions will take string arguments from the expressions, not the values.
+		if e != nil && e.Type() != parser.ValueTypeString {
+			// ev.currentSamples will be updated to the correct value within the ev.eval call.
+			val, ws := ev.eval(e)
+			warnings = append(warnings, ws...)
+			matrixes[i] = val.(Matrix)
+
+			// Keep a copy of the original point slices so that they
+			// can be returned to the pool.
+			origMatrixes[i] = make(Matrix, len(matrixes[i]))
+			copy(origMatrixes[i], matrixes[i])
+		}
+	}
+
+	vectors := make([]Vector, len(exprs)) // Input vectors for the function.
+	// Create an output vector that is as big as the input matrix with
+	// the most time series.
+	biggestLen := 1
+	for i := range exprs {
+		vectors[i] = make(Vector, 0, len(matrixes[i]))
+		if len(matrixes[i]) > biggestLen {
+			biggestLen = len(matrixes[i])
+		}
+	}
+	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
+
+	// The aggregation expects only 1 expression.
+	if len(exprs) != 1 {
+		ev.error(fmt.Errorf("unexpected number of expressions (actual: %d)", len(exprs)))
+	}
+
+	lb := labels.NewBuilder(nil)
+	var buf []byte
+
+	for _, series := range matrixes[0] {
```

Comment:
The idea of this experiment is that instead of iterating on steps (like `rangeEval()`) we iterate on series first and, for each series, we process all steps.